File: app/src/main/java/com/example/rooster/professionaltools/diagnosisassistant/DiagnosisAssistantScreen.kt
package com.example.rooster.professionaltools.diagnosisassistant

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update

// --- Data Classes ---
data class PotentialCondition(
    val condition: String,
    val confidence: Double,
    val urgency: String,
    val nextSteps: String,
    val matchedSymptoms: List<String>
)

data class DiagnosticSessionResult(
    val sessionId: String,
    val species: String,
    val primarySymptoms: List<String>,
    val potentialConditions: List<PotentialCondition>,
    val timestamp: Long,
    val error: String? = null
)

data class ConditionDetails(
    val name: String,
    val information: String
)

// --- ViewModel ---
class DiagnosisAssistantViewModel : ViewModel() {
    private val _sessionResult = MutableStateFlow<DiagnosticSessionResult?>(null)
    val sessionResult: StateFlow<DiagnosticSessionResult?> = _sessionResult

    private val _conditionDetails = MutableStateFlow<ConditionDetails?>(null)
    val conditionDetails: StateFlow<ConditionDetails?> = _conditionDetails

    // Simplified Knowledge Base (mirrors Python structure)
    private val knowledgeBase: Map<String, Map<String, Any>> = mapOf(
        "Cow" to mapOf(
            "symptoms" to mapOf(
                "coughing" to listOf(
                    mapOf("condition" to "Respiratory Infection (BRD Complex)", "confidence" to 0.7, "urgency" to "High", "next_steps" to "Isolate, check temp, consult vet for antibiotics/anti-inflammatory. Good ventilation."),
                    mapOf("condition" to "Lungworm", "confidence" to 0.4, "urgency" to "Medium", "next_steps" to "Fecal test, consult vet for dewormer. Pasture hygiene.")
                ),
                "lameness" to listOf(
                    mapOf("condition" to "Foot Rot", "confidence" to 0.6, "urgency" to "Medium", "next_steps" to "Clean foot, topical treatment, dry footing. Consult vet if severe."),
                    mapOf("condition" to "Injury (Sprain/Fracture)", "confidence" to 0.3, "urgency" to "High", "next_steps" to "Immobilize, consult vet immediately.")
                )
            ),
            "common_conditions_info" to mapOf(
                "Respiratory Infection (BRD Complex)" to "Common in stressed/young cattle. Bacterial/viral. Signs: cough, fever, nasal discharge.",
                "Foot Rot" to "Bacterial infection of the foot, causes severe lameness."
            )
        ),
        "Dog" to mapOf(
            "symptoms" to mapOf(
                "vomiting" to listOf(
                    mapOf("condition" to "Dietary Indiscretion", "confidence" to 0.7, "urgency" to "Low-Medium", "next_steps" to "Withhold food 12-24h, then bland diet. Vet if persistent."),
                    mapOf("condition" to "Gastroenteritis", "confidence" to 0.5, "urgency" to "Medium", "next_steps" to "Consult vet for supportive care."),
                    mapOf("condition" to "Foreign Body Obstruction", "confidence" to 0.3, "urgency" to "Critical", "next_steps" to "IMMEDIATE vet attention if suspected.")
                ),
                "itching_scratching" to listOf(
                    mapOf("condition" to "Flea Allergy Dermatitis (FAD)", "confidence" to 0.7, "urgency" to "Medium", "next_steps" to "Strict flea control. Consult vet for relief."),
                    mapOf("condition" to "Atopic Dermatitis", "confidence" to 0.6, "urgency" to "Medium", "next_steps" to "Consult vet for diagnosis and management plan.")
                )
            ),
            "common_conditions_info" to mapOf(
                "Dietary Indiscretion" to "Caused by eating inappropriate items. Usually self-limiting.",
                "Flea Allergy Dermatitis (FAD)" to "Allergic reaction to flea saliva. Intense itching."
            )
        )
    )

    val availableSpecies: List<String> = knowledgeBase.keys.toList()
    val availableSymptomsBySpecies: Map<String, List<String>> = knowledgeBase.mapValues { entry ->
        ((entry.value["symptoms"] as? Map<String, List<Any>>)?.keys?.toList() ?: emptyList())
    }


    @Suppress("UNCHECKED_CAST")
    fun startDiagnosticSession(species: String, symptoms: List<String>) {
        val speciesKb = knowledgeBase[species]
        if (speciesKb == null) {
            _sessionResult.value = DiagnosticSessionResult("error_${System.currentTimeMillis()}", species, symptoms, emptyList(), System.currentTimeMillis(), "Knowledge base for $species not available.")
            return
        }

        val symptomMap = speciesKb["symptoms"] as? Map<String, List<Map<String, Any>>> ?: emptyMap()
        val possibleMatches = mutableListOf<PotentialCondition>()

        symptoms.forEach { symptomKey ->
            symptomMap[symptomKey]?.forEach { conditionInfo ->
                val conditionName = conditionInfo["condition"] as String
                val existingMatchIndex = possibleMatches.indexOfFirst { it.condition == conditionName }

                if (existingMatchIndex != -1) {
                    val existing = possibleMatches[existingMatchIndex]
                    possibleMatches[existingMatchIndex] = existing.copy(
                        confidence = minOf(1.0, existing.confidence + 0.1 * (conditionInfo["confidence"] as Double)),
                        matchedSymptoms = (existing.matchedSymptoms + symptomKey).distinct()
                    )
                } else {
                    possibleMatches.add(
                        PotentialCondition(
                            condition = conditionName,
                            confidence = conditionInfo["confidence"] as Double,
                            urgency = conditionInfo["urgency"] as String,
                            nextSteps = conditionInfo["next_steps"] as String,
                            matchedSymptoms = listOf(symptomKey)
                        )
                    )
                }
            }
        }
        possibleMatches.sortByDescending { it.confidence }
        _sessionResult.value = DiagnosticSessionResult("diag_${System.currentTimeMillis()}", species, symptoms, possibleMatches, System.currentTimeMillis())
        _conditionDetails.value = null // Clear previous details
    }

    @Suppress("UNCHECKED_CAST")
    fun getConditionDetails(species: String, conditionName: String) {
        val info = (knowledgeBase[species]?.get("common_conditions_info") as? Map<String, String>)?.get(conditionName)
        _conditionDetails.value = ConditionDetails(conditionName, info ?: "No detailed information available.")
    }

    fun clearSession() {
        _sessionResult.value = null
        _conditionDetails.value = null
    }
}

// --- Composable Screen ---
@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun DiagnosisAssistantScreen(viewModel: DiagnosisAssistantViewModel = androidx.lifecycle.viewmodel.compose.viewModel()) {
    val sessionResult by viewModel.sessionResult.collectAsState()
    val conditionDetails by viewModel.conditionDetails.collectAsState()

    var selectedSpecies by remember { mutableStateOf(viewModel.availableSpecies.firstOrNull() ?: "") }
    var symptomsInput by remember { mutableStateOf("") } // Comma-separated
    val keyboardController = LocalSoftwareKeyboardController.current

    Scaffold(
        topBar = {
            TopAppBar(title = { Text("Diagnosis Assistant") }, colors = TopAppBarDefaults.topAppBarColors(containerColor = Color(0xFF5C6BC0))) // Indigo
        }
    ) { paddingValues ->
        Column(modifier = Modifier.padding(paddingValues).padding(16.dp).fillMaxSize()) {
            Text("Disclaimer: This is a mock assistant and NOT a substitute for professional veterinary diagnosis.", style = MaterialTheme.typography.bodySmall, color = Color.Gray)
            Spacer(Modifier.height(8.dp))

            // Inputs
            ExposedDropdownMenuBoxForOptions("Select Species", viewModel.availableSpecies, selectedSpecies) { selectedSpecies = it }
            Spacer(Modifier.height(8.dp))
            OutlinedTextField(
                value = symptomsInput,
                onValueChange = { symptomsInput = it },
                label = { Text("Enter Symptoms (comma-separated)") },
                modifier = Modifier.fillMaxWidth(),
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                keyboardActions = KeyboardActions(onSearch = {
                    val symptomsList = symptomsInput.split(",").map { it.trim() }.filter { it.isNotBlank() }
                    if (selectedSpecies.isNotBlank() && symptomsList.isNotEmpty()) {
                        viewModel.startDiagnosticSession(selectedSpecies, symptomsList)
                    }
                    keyboardController?.hide()
                })
            )
            Spacer(Modifier.height(8.dp))
            Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                 Button(onClick = {
                    val symptomsList = symptomsInput.split(",").map { it.trim() }.filter { it.isNotBlank() }
                    if (selectedSpecies.isNotBlank() && symptomsList.isNotEmpty()) {
                        viewModel.startDiagnosticSession(selectedSpecies, symptomsList)
                    }
                    keyboardController?.hide()
                }) { Icon(Icons.Filled.Search, "Diagnose"); Spacer(Modifier.width(4.dp)); Text("Get Diagnosis") }
                Button(onClick = { viewModel.clearSession(); symptomsInput = "" }, colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.secondary)) { Text("Clear") }
            }


            Spacer(Modifier.height(16.dp))

            // Results
            sessionResult?.let { result ->
                if (result.error != null) {
                    Text("Error: ${result.error}", color = MaterialTheme.colorScheme.error)
                } else {
                    Text("Results for ${result.species} with symptoms: ${result.primarySymptoms.joinToString()}", style = MaterialTheme.typography.titleMedium)
                    if (result.potentialConditions.isEmpty()) {
                        Text("No potential conditions found for the given symptoms.")
                    } else {
                        LazyColumn(modifier = Modifier.weight(1f)) {
                            items(result.potentialConditions) { pc ->
                                PotentialConditionCard(pc) {
                                    viewModel.getConditionDetails(result.species, pc.condition)
                                }
                            }
                        }
                    }
                }
            }

            conditionDetails?.let { details ->
                ConditionDetailsDialog(details = details, onDismiss = { viewModel._conditionDetails.value = null /* Hacky way to clear for now */ })
            }
        }
    }
}

@Composable
fun PotentialConditionCard(pc: PotentialCondition, onViewDetails: () -> Unit) {
    Card(modifier = Modifier.fillMaxWidth().padding(vertical = 6.dp).clickable(onClick = onViewDetails), elevation = CardDefaults.cardElevation(2.dp)) {
        Column(Modifier.padding(12.dp)) {
            Text("${pc.condition} (Confidence: %.2f)".format(pc.confidence), style = MaterialTheme.typography.titleMedium)
            Text("Urgency: ${pc.urgency}", color = when(pc.urgency){ "High" -> Color.Red; "Critical" -> Color.Red; "Medium" -> Color(0xFFFFA000); else -> Color.Green })
            Text("Matched Symptoms: ${pc.matchedSymptoms.joinToString()}", style = MaterialTheme.typography.bodySmall)
            Spacer(Modifier.height(4.dp))
            Text("Next Steps: ${pc.nextSteps}", style = MaterialTheme.typography.bodyMedium)
        }
    }
}

@Composable
fun ConditionDetailsDialog(details: ConditionDetails, onDismiss: () -> Unit) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(details.name) },
        text = { Text(details.information) },
        confirmButton = { Button(onClick = onDismiss) { Text("Close") } }
    )
}

// Re-using ExposedDropdownMenuForOptions
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ExposedDropdownMenuForOptions(
    label: String, options: List<String>, selectedOption: String, onOptionSelected: (String) -> Unit
) {
    var expanded by remember { mutableStateOf(false) }
    ExposedDropdownMenuBox(expanded = expanded, onExpandedChange = { expanded = !expanded }) {
        OutlinedTextField(
            value = selectedOption, onValueChange = {}, readOnly = true, label = { Text(label) },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            modifier = Modifier.menuAnchor().fillMaxWidth()
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            options.forEach { option ->
                DropdownMenuItem(text = { Text(option) }, onClick = { onOptionSelected(option); expanded = false })
            }
        }
    }
}

@Preview(showBackground = true, heightDp = 800)
@Composable
fun PreviewDiagnosisAssistantScreen() {
    MaterialTheme {
        DiagnosisAssistantScreen(viewModel = DiagnosisAssistantViewModel())
    }
}

@Preview(showBackground = true, heightDp = 800)
@Composable
fun PreviewDiagnosisAssistantScreen_WithResults() {
    val viewModel = DiagnosisAssistantViewModel()
    // Simulate a session for preview
    LaunchedEffect(Unit) {
        viewModel.startDiagnosticSession("Dog", listOf("vomiting", "lethargy"))
    }
    MaterialTheme {
        DiagnosisAssistantScreen(viewModel = viewModel)
    }
}

File: app/professional_tools/diagnosis_assistant_screen.py
import datetime
import random

class DiagnosisAssistantScreen:
    def __init__(self):
        # This is a very simplified knowledge base. A real system would be vastly more complex.
        self.knowledge_base = {
            "Cow": {
                "symptoms": {
                    "coughing": [
                        {"condition": "Respiratory Infection (BRD Complex)", "confidence": 0.7, "urgency": "High", "next_steps": "Isolate animal, check temperature, consult vet for antibiotics/anti-inflammatory. Ensure good ventilation."},
                        {"condition": "Lungworm", "confidence": 0.4, "urgency": "Medium", "next_steps": "Consider fecal test, consult vet for dewormer. Manage pasture hygiene."},
                        {"condition": "Allergic Bronchitis (rare)", "confidence": 0.2, "urgency": "Low", "next_steps": "Identify and remove potential allergens. Consult vet if persistent."}
                    ],
                    "lameness": [
                        {"condition": "Foot Rot", "confidence": 0.6, "urgency": "Medium", "next_steps": "Clean and inspect foot, apply topical treatment, ensure dry footing. Consult vet if severe."},
                        {"condition": "Digital Dermatitis", "confidence": 0.5, "urgency": "Medium", "next_steps": "Regular footbaths, topical antibiotics. Consult vet for herd management plan."},
                        {"condition": "Injury (Sprain/Fracture)", "confidence": 0.3, "urgency": "High", "next_steps": "Immobilize if possible, consult vet immediately for diagnosis and treatment."}
                    ],
                    "reduced_milk_yield": [
                        {"condition": "Mastitis", "confidence": 0.8, "urgency": "High", "next_steps": "Perform CMT, check for udder inflammation, consult vet for culture and treatment. Review milking procedures."},
                        {"condition": "Ketosis", "confidence": 0.5, "urgency": "Medium", "next_steps": "Check urine/milk for ketones, provide oral glucose precursors. Adjust diet. Consult vet."},
                        {"condition": "General Stress/Illness", "confidence": 0.3, "urgency": "Varies", "next_steps": "Full clinical exam needed to identify underlying cause. Monitor other symptoms."}
                    ],
                    "diarrhea": [
                        {"condition": "Scours (Calves)", "confidence": 0.7, "urgency": "High", "next_steps": "Oral rehydration therapy, isolate calf, ensure hygiene. Consult vet, especially if bloody."},
                        {"condition": "Dietary Indiscretion/Change", "confidence": 0.4, "urgency": "Low-Medium", "next_steps": "Review feed, gradual diet changes. Provide probiotics. Consult vet if severe or prolonged."},
                        {"condition": "Parasitism (e.g., Coccidiosis)", "confidence": 0.5, "urgency": "Medium", "next_steps": "Fecal exam, consult vet for appropriate dewormer/coccidiostat."}
                    ]
                },
                "common_conditions_info": { # Brief info for when a condition is selected
                    "Respiratory Infection (BRD Complex)": "Common in stressed or young cattle. Bacterial/viral. Key signs: cough, fever, nasal discharge.",
                    "Mastitis": "Udder inflammation, usually bacterial. Reduces milk quality and yield. Key signs: clots in milk, swollen/hot udder.",
                }
            },
            "Dog": {
                "symptoms": {
                    "vomiting": [
                        {"condition": "Dietary Indiscretion", "confidence": 0.7, "urgency": "Low-Medium", "next_steps": "Withhold food for 12-24h, then bland diet. Consult vet if persistent, blood, or other signs."},
                        {"condition": "Gastroenteritis", "confidence": 0.5, "urgency": "Medium", "next_steps": "Consult vet for diagnosis and supportive care. May need fluids/meds."},
                        {"condition": "Foreign Body Obstruction", "confidence": 0.3, "urgency": "Critical", "next_steps": "IMMEDIATE vet attention if suspected (e.g., known ingestion, repeated unproductive vomiting)."}
                    ],
                    "lethargy": [
                        {"condition": "Various Infections (Viral/Bacterial)", "confidence": 0.6, "urgency": "Medium-High", "next_steps": "Check temperature, other symptoms. Consult vet for diagnosis."},
                        {"condition": "Pain (e.g., Arthritis, Injury)", "confidence": 0.4, "urgency": "Medium", "next_steps": "Observe for lameness or pain response. Consult vet."},
                        {"condition": "Systemic Disease (e.g., kidney, heart)", "confidence": 0.3, "urgency": "High", "next_steps": "Consult vet for diagnostics (bloodwork, imaging)."}
                    ],
                    "itching_scratching": [
                        {"condition": "Flea Allergy Dermatitis (FAD)", "confidence": 0.7, "urgency": "Medium", "next_steps": "Strict flea control on pet and environment. Consult vet for relief medication."},
                        {"condition": "Atopic Dermatitis (Environmental Allergies)", "confidence": 0.6, "urgency": "Medium", "next_steps": "Consult vet for diagnosis, allergy testing, and management plan (meds, diet, immunotherapy)."},
                        {"condition": "Food Allergy", "confidence": 0.4, "urgency": "Medium", "next_steps": "Consult vet for diet elimination trial."},
                        {"condition": "Sarcoptic Mange", "confidence": 0.3, "urgency": "High (Contagious)", "next_steps": "Consult vet for skin scrape and treatment. Isolate from other animals."}
                    ]
                },
                 "common_conditions_info": {
                    "Dietary Indiscretion": "Caused by eating inappropriate food items. Usually self-limiting.",
                    "Flea Allergy Dermatitis (FAD)": "Allergic reaction to flea saliva. Intense itching, often around tail base.",
                }
            }
            # Can add more species like Horse, Cat, Poultry, Swine etc.
        }
        self.current_session = None # To store inputs for a single diagnostic query

    def start_diagnostic_session(self, species, primary_symptoms, vet_id="vet_user"):
        if species not in self.knowledge_base:
            print(f"Warning: No detailed knowledge base for species: {species}. Results may be limited.")
            # Could offer generic advice or try to match symptoms across species with lower confidence
            return {"error": f"Knowledge base for {species} not available."}

        self.current_session = {
            "session_id": f"diag_{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}",
            "vet_id": vet_id,
            "species": species,
            "primary_symptoms": primary_symptoms if isinstance(primary_symptoms, list) else [primary_symptoms],
            "potential_conditions": [],
            "timestamp": datetime.datetime.now()
        }

        species_kb = self.knowledge_base[species]["symptoms"]
        possible_matches = []

        for symptom in self.current_session["primary_symptoms"]:
            if symptom in species_kb:
                for condition_info in species_kb[symptom]:
                    # Check if this condition is already added from another symptom to adjust confidence
                    existing_match = next((m for m in possible_matches if m["condition"] == condition_info["condition"]), None)
                    if existing_match:
                        # Increase confidence if multiple symptoms point to same condition (very basic logic)
                        existing_match["confidence"] = min(1.0, existing_match["confidence"] + 0.1 * condition_info["confidence"])
                        existing_match["matched_symptoms"].append(symptom)
                    else:
                        match = condition_info.copy()
                        match["matched_symptoms"] = [symptom]
                        possible_matches.append(match)

        # Sort by confidence (higher first)
        possible_matches.sort(key=lambda x: x["confidence"], reverse=True)
        self.current_session["potential_conditions"] = possible_matches

        return self.current_session

    def get_condition_details(self, species, condition_name):
        if species in self.knowledge_base and "common_conditions_info" in self.knowledge_base[species]:
            return self.knowledge_base[species]["common_conditions_info"].get(condition_name, "No detailed information available for this condition.")
        return "Species or condition information not found."

    def display_screen(self):
        print("---- Diagnosis Assistant Screen ----")
        print("Disclaimer: This is a mock assistant and NOT a substitute for professional veterinary diagnosis.\n")

        # Example 1: Cow with coughing
        print("-- Example 1: Cow with Coughing --")
        session1_symptoms = ["coughing"]
        session1_species = "Cow"
        result1 = self.start_diagnostic_session(session1_species, session1_symptoms)

        if "error" in result1:
            print(f"  Error: {result1['error']}")
        elif result1 and result1["potential_conditions"]:
            print(f"  Species: {session1_species}, Symptoms: {', '.join(session1_symptoms)}")
            print("  Potential Conditions (Top 3):")
            for i, p_cond in enumerate(result1["potential_conditions"][:3]):
                print(f"    {i+1}. {p_cond['condition']} (Confidence: {p_cond['confidence']:.2f}, Urgency: {p_cond['urgency']})")
                print(f"       Matched on: {', '.join(p_cond['matched_symptoms'])}")
                print(f"       Next Steps: {p_cond['next_steps']}")
                # Display further info if available
                # print(f"       More Info: {self.get_condition_details(session1_species, p_cond['condition'])}")
        else:
            print("  No potential conditions found for the given symptoms.")

        # Example 2: Dog with itching and lethargy
        print("\n-- Example 2: Dog with Itching & Lethargy --")
        session2_symptoms = ["itching_scratching", "lethargy"]
        session2_species = "Dog"
        result2 = self.start_diagnostic_session(session2_species, session2_symptoms)

        if "error" in result2:
            print(f"  Error: {result2['error']}")
        elif result2 and result2["potential_conditions"]:
            print(f"  Species: {session2_species}, Symptoms: {', '.join(session2_symptoms)}")
            print("  Potential Conditions (Top 3):")
            for i, p_cond in enumerate(result2["potential_conditions"][:3]):
                print(f"    {i+1}. {p_cond['condition']} (Confidence: {p_cond['confidence']:.2f}, Urgency: {p_cond['urgency']})")
                print(f"       Matched on: {', '.join(p_cond['matched_symptoms'])}")
                print(f"       Next Steps: {p_cond['next_steps']}")
                # Display further info for the top condition
                if i == 0:
                    print(f"       More Info on '{p_cond['condition']}': {self.get_condition_details(session2_species, p_cond['condition'])}")
        else:
            print("  No potential conditions found for the given symptoms.")

        # Example 3: Unknown species (to show error handling)
        print("\n-- Example 3: Unknown Species --")
        session3_symptoms = ["coughing"]
        session3_species = "Sheep" # Assuming Sheep is not in KB
        result3 = self.start_diagnostic_session(session3_species, session3_symptoms)
        if "error" in result3:
            print(f"  Species: {session3_species}, Symptoms: {', '.join(session3_symptoms)}")
            print(f"  Result: {result3['error']}")


        print("----------------------------------")

if __name__ == '__main__':
    screen = DiagnosisAssistantScreen()
    screen.display_screen()
